keras/applications/densenet.py

```diff
@@ -0,0 +1,333 @@
+# -*- coding: utf-8 -*-
+"""DenseNet models for Keras.
+
+# Reference paper:
+
+- [Densely Connected Convolutional Networks](https://arxiv.org/abs/1608.06993) (CVPR 2017 Best Paper Award)
+
+# Reference implementation:
+
+- [Torch DenseNets](https://github.com/liuzhuang13/DenseNet/blob/master/models/densenet.lua)
+- [TensorNets](https://github.com/taehoonlee/tensornets/blob/master/tensornets/densenets.py)
+"""
+from __future__ import print_function
+from __future__ import absolute_import
+
+import os
+
+from .. import backend as K
+from ..models import Model
+from ..layers import Activation
+from ..layers import AveragePooling2D
+from ..layers import BatchNormalization
+from ..layers import Concatenate
+from ..layers import Conv2D
+from ..layers import Dense
+from ..layers import Flatten
+from ..layers import GlobalAveragePooling2D
+from ..layers import Input
+from ..layers import MaxPooling2D
+from ..layers import ZeroPadding2D
+from ..utils.data_utils import get_file
+from ..engine.topology import get_source_inputs
+from .imagenet_utils import decode_predictions
+from .imagenet_utils import _obtain_input_shape
+
+
+__url_prefix__ = 'https://github.com/taehoonlee/deep-learning-models/' \
+                 'releases/download/densenet/'
+
+
+def dense(x, blocks, scope):
+    """A dense block.
+
+    # Arguments
+        x: input tensor.
+        blocks: integer, the number of building blocks.
+        scope: string, block label.
+
+    # Returns
+        output tensor for the block.
+    """
+    for i in range(blocks):
+        x = block(x, 32, scope="%s/block%d" % (scope, i + 1))
+    return x
+
+
+def transition(x, reduction, scope):
+    """A transition block.
+
+    # Arguments
+        x: input tensor.
+        reduction: float, compression rate at transition layers.
+        scope: string, block label.
+
+    # Returns
+        output tensor for the block.
+    """
+    bn_axis = 3 if K.image_data_format() == 'channels_last' else 1
+    x = BatchNormalization(axis=bn_axis, epsilon=1e-5,
+                           name="%s/bn" % scope)(x)
+    x = Activation('relu', name="%s/relu" % scope)(x)
+    x = Conv2D(int(x._keras_shape[bn_axis] * reduction), 1, use_bias=False,
+               name="%s/conv" % scope)(x)
+    x = AveragePooling2D(2, strides=2, name="%s/pool" % scope)(x)
+    return x
+
+
+def block(x, growth_rate, scope):
+    """A building block for a dense block.
+
+    # Arguments
+        x: input tensor.
+        growth_rate: float, growth rate at dense layers.
+        scope: string, block label.
+
+    # Returns
+        output tensor for the block.
+    """
+    bn_axis = 3 if K.image_data_format() == 'channels_last' else 1
+    x1 = BatchNormalization(axis=bn_axis, epsilon=1e-5,
+                            name="%s/0/bn" % scope)(x)
```

Comment:
Use `'` as quote character

Don't use `/` in layer names as this is disallowed in TF (and will be overwritten with something else, maybe `_`)

Reply:
Updated.